src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1718,7 +1718,9 @@ class ShenandoahGenerationalEvacuationTask : public WorkerTask {
           // doing this work during a safepoint.  We cannot put humongous regions into the collection set because that
           // triggers the load-reference barrier (LRB) to copy on reference fetch.
           r->promote_humongous();
-        } else if (r->is_regular() && (r->garbage_before_padded_for_promote() < old_garbage_threshold) && (r->get_top_before_promote() == tams)) {
+        } else if (r->is_regular() && (r->get_top_before_promote() != nullptr)
+                  && (r->garbage_before_padded_for_promote() < old_garbage_threshold)
```

Comment:
I think my earlier guidance may have been wrong here.  I think we can rewrite this as:
else if (r->is_regular() && (r->get_top_before_promote() != nullptr)) {
  assert(r->garbage_before_padded_for_promote() < old_garbage_threshold, ...)
  assert (r->get_top_before_promote() == tams, ...)

Reply:
Okay, will make that change.